pkg/util/chunk/chunk_in_disk.go

```diff
@@ -0,0 +1,240 @@
+// Warning: Do not concurrently call this function.
+func (d *DataInDiskByChunks) Add(chk *Chunk) (err error) {
+	if chk.NumRows() == 0 {
+		return errors2.New("Chunk spilled to disk should have at least 1 row")
+	}
+
+	if d.dataFile.file == nil {
+		err = d.initDiskFile()
+		if err != nil {
+			return
+		}
+	}
+
+	serializedBytesNum := d.serializeDataToBuf(chk)
+
+	var writeNum int
+	writeNum, err = d.dataFile.write(d.buf)
+	if err != nil {
+		return
+	}
+
+	if int64(writeNum) != serializedBytesNum {
+		return errors2.New("Some data fail to be spilled to disk")
+	}
+	d.offsetOfEachChunk = append(d.offsetOfEachChunk, d.totalDataSize)
+	d.totalDataSize += serializedBytesNum
+	d.totalRowNum += int64(chk.NumRows())
+	d.dataFile.offWrite += serializedBytesNum
+
+	d.diskTracker.Consume(serializedBytesNum)
+	return
+}
+
+func (d *DataInDiskByChunks) getChunkSize(chkIdx int) int64 {
+	totalChunkNum := len(d.offsetOfEachChunk)
+	if chkIdx == totalChunkNum-1 {
+		return d.totalDataSize - d.offsetOfEachChunk[chkIdx]
+	}
+	return d.offsetOfEachChunk[chkIdx+1] - d.offsetOfEachChunk[chkIdx]
+}
+
+// GetChunk gets a Chunk from the DataInDiskByChunks by chkIdx.
+func (d *DataInDiskByChunks) GetChunk(chkIdx int) (*Chunk, error) {
+	reader := bufio.NewReader(d.dataFile.getSectionReader(d.offsetOfEachChunk[chkIdx]))
+	chkSize := d.getChunkSize(chkIdx)
+
+	if len(d.buf) < int(chkSize) {
+		d.buf = make([]byte, chkSize)
+	} else {
+		d.buf = d.buf[:chkSize]
+	}
+
+	readByteNum, err := io.ReadFull(reader, d.buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if int64(readByteNum) != chkSize {
+		return nil, errors2.New("Fail to restore the spilled chunk")
+	}
+
+	chk := NewChunkWithCapacity(d.fieldTypes, 2048)
+	d.deserializeDataToChunk(d.buf, chk)
+
+	return chk, nil
+}
+
+// Close releases the disk resource.
+func (d *DataInDiskByChunks) Close() error {
+	if d.dataFile.file != nil {
+		d.diskTracker.Consume(-d.diskTracker.BytesConsumed())
+		terror.Call(d.dataFile.file.Close)
+		terror.Log(os.Remove(d.dataFile.file.Name()))
+	}
+	return nil
+}
+
+// Format of the chunk serialized in buffer:
+// row1: | col1 size | col1 data | col2 size | col2 data | col3 size | col3 data |
+// row2: | col1 size | col1 data | col2 size | col2 data | col3 size | col3 data |
+// row3: | col1 size | col1 data | col2 size | col2 data | col3 size | col3 data |
+// row4: | col1 size | col1 data | col2 size | col2 data | col3 size | col3 data |
+//
+// Column size will be -1 if column data is null.
+func (d *DataInDiskByChunks) serializeDataToBuf(chk *Chunk) int64 {
+	d.buf = d.buf[0:0] // clear the buf data
+
+	rowNum := chk.NumRows()
+	colNum := chk.NumCols()
+	colSizeRecordBytes := int64(colSizeMetaLen * colNum)
+	totalBytes := int64(0)
+	addedBytesNum := int64(0)
+	rowBuf := make([]byte, 0, 1024)
+	tmpBuf := make([]byte, colSizeMetaLen) // Used for storing column size
+
+	for i := 0; i < rowNum; i++ {
```

Comment:
Why not serialize by column, just like tiflash?

Reply:
Column-wise serialization and deserialization should have better performance.

Reply:
> Column-wise serialization and deserialization should have better performance.

done

Reply:
> Column-wise serialization and deserialization should have better performance.

After benchmark, we increase performance by 38.8%